apple/internal/codesigning_support.bzl

```diff
@@ -422,6 +426,91 @@ def _codesigning_command(
         codesignopts = codesignopts,
     )
 
+def _generate_codesigning_dossier_action(
+        actions,
+        label_name,
+        resolved_codesigning_dossier_tool,
+        output_dossier,
+        platform_prerequisites,
+        embedded_dossiers = [],
+        entitlements = None,
+        provisioning_profile = None):
+    """Generates a codesigning dossier based on parameters.
+
+    Args:
+      actions: The actions provider from `ctx.actions`.
+      embedded_dossiers: An optional List of Structs generated from
+         `embedded_codesigning_dossier` that should also be included in this
+          dossier.
+      entitlements: Optional file representing the entitlements to sign with.
+      label_name: Name of the target being built.
+      output_dossier: The `File` representing the output dossier file - the zipped dossier will be placed here.
+      platform_prerequisites: Struct containing information on the platform being targeted.
+      provisioning_profile: The provisioning profile file. May be `None`.
+      resolved_codesigning_dossier_tool: The `struct` from resolve_tools representing the code signing tool.
+    """
+    input_files = [x.dossier_file for x in embedded_dossiers]
+
+    mnemonic = "GenerateCodesigningDossier"
+    progress_message = "Generating codesigning dossier for %s" % label_name
+
+    dossier_arguments = ["--output", output_dossier.path, "--zip"]
+
+    is_device = platform_prerequisites.platform.is_device
+    fragment = platform_prerequisites.objc_fragment
+    codesign_identity = fragment.signing_certificate_name if is_device else "-"
+    if not codesign_identity and not provisioning_profile:
+        codesign_identity = "-"
+    if codesign_identity:
+        dossier_arguments.extend(["--codesign_identity", codesign_identity])
+    if entitlements:
+        input_files.append(entitlements)
+        dossier_arguments.extend(["--entitlements_file", entitlements.path])
+    if provisioning_profile:
+        input_files.append(provisioning_profile)
+        dossier_arguments.extend(["--provisioning_profile", provisioning_profile.path])
+
+    for embedded_dossier in embedded_dossiers:
+        input_files.append(embedded_dossier.dossier_file)
+        dossier_arguments.extend(["--embedded_dossier", embedded_dossier.relative_bundle_path, embedded_dossier.dossier_file.path])
+
+    args_file = intermediates.file(
```

Comment:
this feels like it should be doable with an `actions.args()`

Reply:
This would be a nice improvement, but since the open-source version of the rules is not using this code path, I'm leaving it as-is to help keep things in sync.